cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,137 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of queue.
+ *
+ * Implemented methods are not boxed, when using with primitive types.
+ */
+abstract class CQueue[@specialized(Int, Long) T] {
+  def first(): T
+  def +=(elem: T)
+  def deque(): T
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]].
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  def enqueueFirst(elem: T)
+}
+
+object CQueue {
```

Comment:
I think if this extends CQueueFactory, you can just delegate to it rather than using impls. Like how object Map extends MapFactory (in scala.collection.Map)